File: src/csharp/pulse/cdm/scalars/SEScalarVolumePerTimeMass.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class VolumePerTimeMassUnit : Unit
  {
    private VolumePerTimeMassUnit(string v) { Value = v; }

    public static VolumePerTimeMassUnit L_Per_s_g = new VolumePerTimeMassUnit("L/s g");
    public static VolumePerTimeMassUnit mL_Per_s_g = new VolumePerTimeMassUnit("mL/s g");
    public static VolumePerTimeMassUnit mL_Per_min_kg = new VolumePerTimeMassUnit("mL/min kg");
    public static VolumePerTimeMassUnit mL_Per_s_kg = new VolumePerTimeMassUnit("mL/s kg");
    public static VolumePerTimeMassUnit uL_Per_min_kg = new VolumePerTimeMassUnit("uL/min kg");

    public static VolumePerTimeMassUnit FromString(string u)
    {
      if (u == VolumePerTimeMassUnit.L_Per_s_g.ToString())
        return VolumePerTimeMassUnit.L_Per_s_g;
      if (u == VolumePerTimeMassUnit.mL_Per_s_g.ToString())
        return VolumePerTimeMassUnit.mL_Per_s_g;
      if (u == VolumePerTimeMassUnit.mL_Per_min_kg.ToString())
        return VolumePerTimeMassUnit.mL_Per_min_kg;
      if (u == VolumePerTimeMassUnit.mL_Per_s_kg.ToString())
        return VolumePerTimeMassUnit.mL_Per_s_kg;
      if (u == VolumePerTimeMassUnit.uL_Per_min_kg.ToString())
        return VolumePerTimeMassUnit.uL_Per_min_kg;

      throw new System.ArgumentException(u + " is not a know volume per time mass unit");
    }
  }

  public class SEScalarVolumePerTimeMass : SEScalar
  {
    protected VolumePerTimeMassUnit unit;

    public SEScalarVolumePerTimeMass() : base()
    {

    }

    public SEScalarVolumePerTimeMass(SEScalarVolumePerTimeMass from)
    {
      Set(from);
    }

    public SEScalarVolumePerTimeMass(double value, VolumePerTimeMassUnit unit)
    {
      SetValue(value, unit);
    }

    public void Set(SEScalarVolumePerTimeMass from)
    {
      if (from == null)
      {
        Invalidate();
        return;
      }
      SetValue(from.value, from.unit);
    }

    public new void SetValue(double value)
    {
      throw new System.NotSupportedException("You must provide a unit");
    }

    public void SetValue(double value, VolumePerTimeMassUnit unit)
    {
      this.value = value;
      this.unit = unit;
    }

    public double GetValue(VolumePerTimeMassUnit unit)
    {
      return UnitConverter.Convert(this.value, this.unit, unit);
    }

    public VolumePerTimeMassUnit GetUnit()
    {
      return unit;
    }

    public new string ToString()
    {
      return this.value + "(" + this.unit.ToString() + ")";
    }
  }
}

File: src/cpp/cdm/properties/SEScalarVolumePerTimeMass.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarVolumePerTimeMass.h"

const VolumePerTimeMassUnit VolumePerTimeMassUnit::L_Per_s_g("L/s g");
const VolumePerTimeMassUnit VolumePerTimeMassUnit::mL_Per_s_g("mL / s g");
const VolumePerTimeMassUnit VolumePerTimeMassUnit::mL_Per_min_kg("mL/min kg");
const VolumePerTimeMassUnit VolumePerTimeMassUnit::mL_Per_s_kg("mL/s kg");
const VolumePerTimeMassUnit VolumePerTimeMassUnit::uL_Per_min_kg("uL/min kg");

bool VolumePerTimeMassUnit::IsValidUnit(const std::string& unit)
{
  if (L_Per_s_g.GetString().compare(unit) == 0)
    return true;
  if (mL_Per_s_g.GetString().compare(unit) == 0)
    return true;
  if (mL_Per_min_kg.GetString().compare(unit) == 0)
    return true;
  if (mL_Per_s_kg.GetString().compare(unit) == 0)
    return true;
  if (uL_Per_min_kg.GetString().compare(unit) == 0)
    return true;
  return false;
}

const VolumePerTimeMassUnit& VolumePerTimeMassUnit::GetCompoundUnit(const std::string& unit)
{
  if (L_Per_s_g.GetString().compare(unit) == 0)
    return L_Per_s_g;
  if (mL_Per_s_g.GetString().compare(unit) == 0)
    return mL_Per_s_g;
  if (mL_Per_min_kg.GetString().compare(unit) == 0)
    return mL_Per_min_kg;
  if (mL_Per_s_kg.GetString().compare(unit) == 0)
    return mL_Per_s_kg;
  if (uL_Per_min_kg.GetString().compare(unit) == 0)
    return uL_Per_min_kg;
  throw CommonDataModelException(unit + " is not a valid VolumePerTimeMass unit");
}

template class SEScalarQuantity<VolumePerTimeMassUnit>;
